src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -40,24 +39,37 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
         /// but the ngen image will exist for the basic map between two reference types, since those are reused.</remarks>
         private readonly IBidirectionalMap<string, Tuple<Guid>> _languageMap;
 
+        private IVsTextManager4 _textManager;
+        private ComEventSink _textManagerEvents2Sink;
+
         /// <remarks>
         /// We make sure this code is from the UI by asking for all serializers on the UI thread in <see cref="HACK_AbstractCreateServicesOnUiThread"/>.
         /// </remarks>
         [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public LanguageSettingsPersister(
-            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
-            IGlobalOptionService optionService)
+            IGlobalOptionService optionService, [Import(typeof(SAsyncServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
             : base(assertIsForeground: true)
         {
-            _textManager = (IVsTextManager4)serviceProvider.GetService(typeof(SVsTextManager));
             _optionService = optionService;
+            _serviceProvider = serviceProvider;
 
             // TODO: make this configurable
             _languageMap = BidirectionalMap<string, Tuple<Guid>>.Empty.Add(LanguageNames.CSharp, Tuple.Create(Guids.CSharpLanguageServiceId))
                                                                .Add(LanguageNames.VisualBasic, Tuple.Create(Guids.VisualBasicLanguageServiceId))
                                                                .Add("TypeScript", Tuple.Create(new Guid("4a0dddb5-7a95-4fbf-97cc-616d07737a77")))
                                                                .Add("F#", Tuple.Create(new Guid("BC6DD5A5-D4D6-4dab-A00D-A51242DBAF1B")))
                                                                .Add("Xaml", Tuple.Create(new Guid("CD53C9A1-6BC2-412B-BE36-CC715ED8DD41")));
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _textManager = (IVsTextManager4)await _serviceProvider.GetServiceAsync(typeof(SVsTextManager)).ConfigureAwait(true);
```

Comment:
What worries me with this approach is we're still blindly assuming then this will actually run prior to anybody else making the request for the values. Do we have a better approach here?

Reply:
I think the way JTF assumes is, fully embrace JTF and do the initialization when it is actually used and use JTF to convert async call to sync call if needed and do UI switch stuff if needed and etc.

which I am not sure whether we can do without using JTF in every places even in lower layers.

Reply:
It's sort of more a MEF-related rule to be free-threaded in part activation to avoid deadlocks rather than a JTF rule. 
But ya, @jasonmalinowski's concern is certainly valid if folks will import this part and use it without calling `InitializeAsync` on it first. My general recommendation is make this init method safe to call many times (including concurrently) and make sure everyone who imports the MEF export calls the init method first. One common way to safely init considering concurrent callers and possible main thread requirements is to use the vs-threading's `AsyncLazy<T>` class, passing in a JTF instance to its ctor. Then your init call is simply a query for `AsyncLazy<T>.Value` for the sake of its side-effect of running the value factory just once.

Another option is simply to make every other entrypoint into your MEF part run the initialize method first so that all callers will implicitly init it if it isn't already.

Reply:
This particular case is problematic: this interface is internal and so we can change contracts as necessary, but this code can be called indirectly through a public API. That can be done on a background thread where our rule was that is fully free threaded since it's available through our Workspace snapshot model. Initializing it now in a way that requires the UI thread to be either unblocked or blocked in a JTF-friendly away is technically a breaking change.

Combining another conversation that happened offline, the other icky bit is us asking the service for the values is surprisingly expensive once we have the service; we're looking to see if we have other options and can just stop using it entirely. So we might be able to avoid it, but that doesn't mean the conversation here still isn't good!

Reply:
> That can be done on a background thread where our rule was that is fully free threaded since it's available through our Workspace snapshot model. Initializing it now in a way that requires the UI thread to be either unblocked or blocked in a JTF-friendly away is technically a breaking change.

I would agree going from purely free-threaded to requiring the UI thread is a potentially breaking change for folks. But it seems this was never *really* purely free-threaded. I guess it had thread affinity for activation and was free-threaded after that (assuming the `_textManager` isn't used or is free-threaded itself, which I don't know), right? And now that we're dealing with the fact that activation is supposed to be free-threaded to guard against deadlocks, that pushes the thread affinity downstream to your callers. It cuts both ways. ☹️ You may need to make a call based on which way you think will lead to the fewest deadlocks. But just one thought: keeping MEF activation thread affinitized may mean you're occasionally bit by deadlocks forever. But if you move thread affinity outside of MEF, while you may hit deadlocks in 3rd party callers for a while, after they adjust the deadlocks should end.